### PR TITLE
hwspinlock: add hwspinlock driver for multi CPU

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
Can be used in AMP case, the design follow Linux:
https://www.kernel.org/doc/html/latest/locking/hwspinlock.html

## Impact
New api

### CODE REVIEW

File: include/nuttx/hwspinlock/hwspinlock.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+{
+  irqstate_t flags = spin_lock_irqsave(&dev->lock);
+  hwspin_lock(dev, id, priority);
+  return flags;
+}
+
+static inline void hwspin_unlock(FAR struct hwspinlock_dev_s *dev,
+                                 int id, int priority)
+{
+  dev->ops->unlock(dev, id);
```
Comment (pkarashchenko) [inline include/nuttx/hwspinlock/hwspinlock.h:111]: Why we can't call it directly?

File: include/nuttx/hwspinlock/hwspinlock.h [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+extern "C"
+{
+#else
+#define EXTERN extern
+#endif
+
+static inline bool hwspin_trylock(FAR struct hwspinlock_dev_s *dev,
+                                  int id, int priority)
+{
+  return dev->ops->trylock(dev, id, priority);
```
Comment (pkarashchenko) [inline include/nuttx/hwspinlock/hwspinlock.h:70]: Why we can't call it directly?

Comment (xiaoxiang781216) [reply include/nuttx/hwspinlock/hwspinlock.h:70]: simple and unify the usage(e.g. hwspin_trylock v..s. hwspin_trylock_irqsave)

File: include/nuttx/hwspinlock/hwspinlock.h (hunk 1 above)
Comment (xiaoxiang781216) [reply include/nuttx/hwspinlock/hwspinlock.h:111]: ditto

File: include/nuttx/hwspinlock/hwspinlock.h (hunk 2 above)
Comment (pkarashchenko) [reply include/nuttx/hwspinlock/hwspinlock.h:69]: I see. That is a separate API. My initial feeling was that it is only used below, but it is exposed to the user.

**Review by pkarashchenko — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]

**Pull request merged**